runtime/gc_realtime/RealtimeRootScanner.hpp

```diff
@@ -117,5 +118,376 @@ class MM_RealtimeRootScanner : public MM_RootScanner
+	 * @see MM_RootScanner::scanOneThread()
+	 */
+	virtual void
+	scanOneThreadImpl(MM_EnvironmentRealtime *env, J9VMThread* walkThread, void* localData)
+	{
+		MM_EnvironmentRealtime* walkThreadEnv = MM_EnvironmentRealtime::getEnvironment(walkThread->omrVMThread);
+		/* Scan the thread by invoking superclass */
+		MM_RootScanner::scanOneThread(env, walkThread, localData);
+
+		/*
+		 * TODO CRGTMP we should be able to premark the cache instead of flushing the cache
+		 * but this causes problems in overflow.  When we have some time we should look into
+		 * this again.
+		 */
+		/*((MM_SegregatedAllocationInterface *)walkThreadEnv->_objectAllocationInterface)->preMarkCache(walkThreadEnv);*/
+		walkThreadEnv->_objectAllocationInterface->flushCache(walkThreadEnv);
+		/* Disable the double barrier on the scanned thread. */
+		_realtimeGC->disableDoubleBarrierOnThread(env, walkThread->omrVMThread);
+	}
+	
+#if defined(J9VM_GC_FINALIZATION)
+	virtual void doFinalizableObject(j9object_t object) {
+		_markingScheme->markObject(_env, object);
+	}
+#endif /* J9VM_GC_FINALIZATION */
+
+	/**
+	 * Simply pass the call on to the RealtimeGC.
+	 * @see MM_Metronome::markObject()
+	 */
+	virtual void
+	doSlot(J9Object** slot)
+	{
+		_markingScheme->markObject(_env, *slot);
+	}
+	
+	virtual void
+	doStackSlot(J9Object **slotPtr, void *walkState, const void* stackLocation)
+	{
+		J9Object *object = *slotPtr;
+		if (_markingScheme->isHeapObject(object)) {
+			/* heap object - validate and mark */
+			Assert_MM_validStackSlot(MM_StackSlotValidator(0, object, stackLocation, walkState).validate(_env));
+			_markingScheme->markObject(_env, object);
+		} else if (NULL != object) {
+			/* stack object - just validate */
+			Assert_MM_validStackSlot(MM_StackSlotValidator(MM_StackSlotValidator::NOT_ON_HEAP, object, stackLocation, walkState).validate(_env));
+		}
+	}
+
+	virtual void 
+	doVMThreadSlot(J9Object **slotPtr, GC_VMThreadIterator *vmThreadIterator) {
+		J9Object *object = *slotPtr;
+		if (_markingScheme->isHeapObject(object)) {
+			_markingScheme->markObject(_env, object);
+		} else if (NULL != object) {
+			Assert_MM_true(vmthreaditerator_state_monitor_records == vmThreadIterator->getState());
+		}
+	}
+
+	/**
+	 * Scans non-collectable internal objects (immortal) and classes
+	 */
+	virtual void
+	scanIncrementalRoots(MM_EnvironmentRealtime *env)
+	{
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+		if (_classDataAsRoots) {
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+			scanClasses(env);
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+		} else {
+			scanPermanentClasses(env);
+		}
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+
+		CompletePhaseCode status = scanClassesComplete(env);
+
+		if (complete_phase_ABORT == status) {
+			return ;
+		}
+	}
+protected:
+private:
+};
+
+/**
+ * This scanner will mark objects that pass through its doSlot.
+ */
+class MM_RealtimeMarkingSchemeRootClearer : public MM_RealtimeRootScanner
```

Comment:
Would you please create MM_RealtimeMarkingSchemeRootClearer.hpp (and MM_RealtimeMarkingSchemeRootClearer.cpp if necessary) and move this code there?